planner/core/preprocess.go

```diff
@@ -1685,3 +1697,103 @@ func (p *preprocessor) hasAutoConvertWarning(colDef *ast.ColumnDef) bool {
 	}
 	return false
 }
+
+func tryLockMDLAndUpdateSchemaIfNecessary(sctx sessionctx.Context, dbName model.CIStr, tbl table.Table, is infoschema.InfoSchema) (table.Table, error) {
+	if !variable.EnableMDL.Load() {
+		return tbl, nil
+	}
+	if is.SchemaMetaVersion() == 0 {
+		return tbl, nil
+	}
+	skipLock := false
+	if sctx.GetSessionVars().SnapshotInfoschema != nil {
+		return tbl, nil
+	}
+	if sctx.GetSessionVars().TxnCtx.IsStaleness {
+		return tbl, nil
+	}
+	if tbl.Meta().TempTableType == model.TempTableLocal {
+		// Don't attach, don't lock.
+		return tbl, nil
+	} else if tbl.Meta().TempTableType == model.TempTableGlobal {
+		skipLock = true
+	}
+	if IsAutoCommitTxn(sctx) && sctx.GetSessionVars().StmtCtx.IsReadOnly {
+		return tbl, nil
+	}
+	tableInfo := tbl.Meta()
+	if _, ok := sctx.GetSessionVars().GetRelatedTableForMDL().Load(tableInfo.ID); !ok {
+		if se, ok := is.(*infoschema.SessionExtendedInfoSchema); ok && skipLock {
+			if se.MdlTables == nil {
+				return tbl, nil
+			}
+			if _, ok := se.MdlTables.TableByID(tableInfo.ID); ok {
+				// Already attach.
+				return tbl, nil
+			}
+		}
+		var err error
+		tbl, err = domain.GetDomain(sctx).InfoSchema().TableByName(dbName, tableInfo.Name)
+		if err != nil {
+			return nil, err
+		}
+		// Check the table change, if adding new public index or modify a column, we need to handle them.
+		if !sctx.GetSessionVars().IsPessimisticReadConsistency() {
+			var copyTableInfo *model.TableInfo
+			for i, idx := range tbl.Meta().Indices {
+				if idx.State != model.StatePublic {
+					continue
+				}
+				found := false
+				for _, idxx := range tableInfo.Indices {
+					if idx.Name.L == idxx.Name.L && idx.ID == idxx.ID {
+						found = true
+						break
+					}
+				}
+				if !found {
+					if copyTableInfo == nil {
+						copyTableInfo = tbl.Meta().Clone()
+					}
+					copyTableInfo.Indices[i].State = model.StateWriteReorganization
+					dbInfo, _ := domain.GetDomain(sctx).InfoSchema().SchemaByName(dbName)
+					allocs := autoid.NewAllocatorsFromTblInfo(sctx.GetStore(), dbInfo.ID, copyTableInfo)
+					tbl, err = table.TableFromMeta(allocs, copyTableInfo)
+					if err != nil {
+						return nil, err
+					}
+				}
+			}
+			// Check the column change.
+			for _, col := range tbl.Meta().Columns {
+				if col.State != model.StatePublic {
+					continue
+				}
+				found := false
+				for _, coll := range tableInfo.Columns {
+					if col.Name.L == coll.Name.L && col.ID != coll.ID {
+						logutil.BgLogger().Info("public column changed", zap.String("column", col.Name.L), zap.String("old_col", coll.Name.L), zap.Int64("new id", col.ID), zap.Int64("old id", coll.ID))
+						found = true
+						break
+					}
+				}
+				if found {
+					return nil, ErrSchemaChanged.GenWithStack("public column %s has changed", col.Name)
```

Comment:
If here return get mdl lock failed, it will cause txn failed and need user retry txn, right?

Reply:
the transaction is not failed, only the statement failed.